google-cloud-storage/lib/google/cloud/storage/file/signer_v4.rb

```diff
@@ -119,22 +115,22 @@ def issuer_and_signer issuer, client_email, signing_key, private_key
             [issuer, signer]
           end
 
-          def canonical_and_signed_headers headers
-            # Headers needs to be in alpha order.
+          def canonical_and_signed_headers headers, virtual_hosted_style, bucket_bound_hostname
+            if virtual_hosted_style && bucket_bound_hostname
+              raise "virtual_hosted_style: #{virtual_hosted_style} and bucket_bound_hostname: " \
+                    "#{bucket_bound_hostname} params cannot both be passed together"
+            end
+
             canonical_headers = headers || {}
             headers_arr = canonical_headers.map do |k, v|
-              [k.downcase, v.strip.gsub(/[^\S\t]+/, " ").gsub(/\t+/, "\t")]
+              [k.downcase, v.strip.gsub(/[^\S\t]+/, " ").gsub(/\t+/, " ")]
             end
             canonical_headers = Hash[headers_arr]
-            canonical_headers["host"] = "storage.googleapis.com"
+            canonical_headers["host"] = host_name virtual_hosted_style, bucket_bound_hostname
 
-            canonical_headers = canonical_headers.sort_by do |k, _|
-              k.downcase
-            end.to_h
+            canonical_headers = canonical_headers.sort.to_h
             canonical_headers_str = ""
-            canonical_headers.each do |k, v|
-              canonical_headers_str += "#{k}:#{v}\n"
-            end
+            canonical_headers.each { |k, v| canonical_headers_str += "#{k}:#{v}\n" }
```

Comment:
I think it's good practice to avoid mutable strings. Consider:

```ruby
canonical_headers_str = canonical_headers.map { |k, v| "#{k}:#{v}\n" }.join
```

Reply:
Good catch, totally agree.

Reply:
I looked, but I can't find a rubocop cop for this.

Reply:
I don't think there is one.

Probably the closest is the one that requires `# frozen_string_literal: true`. 😈

Reply:
Fixed.